In [47]:
#API imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Step1: Download .csv and Github repo
(yeah, done that)

# Step 2: Data Exploration
1. Load data set
---

In [48]:
#read file
df = pd.read_csv("team11_transport.csv") #this process changes when using collab
df.head()

trip_id        date    route vehicle_type   time_of_day  passenger_count  \
0        1  2023-10-01  Route E          Bus  Morning Rush             39.0   
1        2  2023-03-06  Route E          Bus  Evening Rush             23.0   
2        3  2023-12-09  Route A        Train  Morning Rush             26.0   
3        4  2023-10-20  Route D          Bus        Midday             26.0   
4        5  2023-10-15  Route A          Bus  Morning Rush             33.0   

   delay_minutes fare_type payment_method day_type  
0            0.0   Student           Card  Weekday  
1            0.0     Child     Mobile App  Weekend  
2            NaN     Adult           Card  Weekend  
3            3.0     Adult           Cash  Weekday  
4           12.0   Student           Card  Weekend

2. Explore Data Structure and Types
---

In [49]:
#what type of data are we expecting? 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1300 entries, 0 to 1299
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   trip_id          1300 non-null   int64  
 1   date             1300 non-null   object 
 2   route            1300 non-null   object 
 3   vehicle_type     1300 non-null   object 
 4   time_of_day      1300 non-null   object 
 5   passenger_count  1237 non-null   float64
 6   delay_minutes    1251 non-null   float64
 7   fare_type        1300 non-null   object 
 8   payment_method   1300 non-null   object 
 9   day_type         1300 non-null   object 
dtypes: float64(2), int64(1), object(7)
memory usage: 101.7+ KB


3. Identify data quality issues

In [50]:
#Brief overview of the data at hand
df.describe()

trip_id  passenger_count  delay_minutes
count  1300.000000      1237.000000    1251.000000
mean    650.500000        30.133387       2.549161
std     375.421985         5.491755       2.997296
min       1.000000        15.000000       0.000000
25%     325.750000        26.000000       0.000000
50%     650.500000        30.000000       2.000000
75%     975.250000        34.000000       4.000000
max    1300.000000        49.000000      19.000000

In [51]:
#dimensions of the CSV
df.shape

(1300, 10)

In [52]:
#sum number of null data for all columns
df.isnull().sum()

trip_id             0
date                0
route               0
vehicle_type        0
time_of_day         0
passenger_count    63
delay_minutes      49
fare_type           0
payment_method      0
day_type            0
dtype: int64

In [53]:
for column in df.columns:
    missing = df[column].isna().sum() / df.shape[0]
    print(f"{column:{15}}: -> {missing * 100:.2f}%")

trip_id        : -> 0.00%
date           : -> 0.00%
route          : -> 0.00%
vehicle_type   : -> 0.00%
time_of_day    : -> 0.00%
passenger_count: -> 4.85%
delay_minutes  : -> 3.77%
fare_type      : -> 0.00%
payment_method : -> 0.00%
day_type       : -> 0.00%


That data contains entries with missing data
3.77% of entered delays in minutes are missing from the data.
4.85% of entered passenger count is missing from data

4. Documented findings:
---





In [ ]:
#List all unqiue values in data (This helped us identify inconsistenties and missing values)
for column in df.columns:
    print(f"{column:{15}}: {df[column].unique()}")

trip_id        : [   1    2    3 ... 1298 1299 1300]
date           : ['2023-10-01' '2023-03-06' '2023-12-09' '2023-10-20' '2023-10-15'
 '2023-02-23' '2023-10-31' '2023-07-03' '2023-04-21' '2023-07-27'
 '2023-05-03' '2023-02-07' '2023-11-05' '2023-03-04' '2023-01-01'
 '2023-09-14' '2023-12-19' '2023-05-18' '2023-01-06' '2023-12-31'
 '2023-06-01' '2023-09-22' '2023-06-10' '2023-01-15' '2023-12-12'
 '2023-01-10' '2023-05-26' '2023-01-26' '2023-09-17' '2023-06-18'
 '2023-08-23' '2023-07-19' '2023-12-11' '2023-08-27' '2023-05-13'
 '2023-01-04' '2023-12-29' '2023-01-08' '2023-01-02' '2023-09-01'
 '2023-10-13' '2023-05-15' '2023-10-27' '2023-04-28' '2023-09-20'
 '2023-04-10' '2023-08-28' '2023-03-05' '2023-02-27' '2023-11-13'
 '2023-03-11' '2023-03-31' '2023-10-14' '2023-07-09' '2023-06-17'
 '2023-04-14' '2023-03-07' '2023-04-19' '2023-02-21' '2023-11-14'
 '2023-12-27' '2023-09-10' '2023-10-21' '2023-12-16' '2023-05-05'
 '2023-02-22' '2023-08-07' '2023-02-26' '2023-10-16' '2023-03-27'
 '2023

In [55]:
#account number of each fare type
df['fare_type'].value_counts()

fare_type
Adult      666
Student    295
Senior     184
Child      155
Name: count, dtype: int64

In [56]:
#account number of payment method
df['payment_method'].value_counts()

payment_method
Card          793
Mobile App    271
Cash          236
Name: count, dtype: int64

In [57]:
#account number of day_type
df['day_type'].value_counts()

day_type
Weekday    898
Weekend    402
Name: count, dtype: int64

In [58]:
#account number of vehicle types
df['vehicle_type'].value_counts()

vehicle_type
Bus      662
Train    445
Tram     193
Name: count, dtype: int64

In [59]:
#account number of route types
df['route'].value_counts()

route
Route B    276
Route E    268
Route D    259
Route A    258
Route C    239
Name: count, dtype: int64

In [ ]:
#account number of vehicle types
df['vehicle_type'].value_counts()

In [60]:
#account number of time of day
df['time_of_day'].value_counts()

time_of_day
Evening Rush    449
Morning Rush    442
Midday          292
Night           117
Name: count, dtype: int64

### From 1300 total travel route records:
- The transit data handled 1237 total passengers
- Averaging around 30 passengers per route
- 15 to 49 passengers (3.8% incomplete)
- maximum 19 minute delay (4.8% incomplete)
#### Note:


| Fare Type | # |
| ----------- | ----------- |
| Senior | 184 |
| Adult | 666 |
| Student | 295 |
| Child | 155 |

| Payment Method | # |
| ----------- | ----------- |
| Card | 793 |
| Mobile App | 271 |
| Cash | 236 |

| Day Type | # |
| ----------- | ----------- |
| Weekday | 898 |
| Weekend | 402 |

| Vehicle Type | # |
| ----------- | ----------- |
| Bus | 662 |
| Train | 445 |
| Tram | 193 |

| Routes | # |
| ----------- | ----------- |
| B | 276 |
| E | 268 | 
| D | 259 |
| A | 258 |
| C | 239 |


| Time of Day | # |
| ----------- | ----------- |
| Evening Rush | 449 |
| Morning Rush | 442
| Midday | 292 |
| Night | 117 |


# Step 3: Data Cleaning
---
By checking the number of data the data row and columns. We find that only "Passenger Count" and "Delays" contained inconsistenies. Which are only NaN values


In [ ]:
# get unfiltered shape
df.shape

(1300, 10)

In [ ]:
# .dropna() drops all entries with NaN from the table. (Removes all rows with atleast one NaN) 
dfClean = df.dropna(axis=0)

In [ ]:
#notice how a few rows have been dropped?
dfClean.shape

(1188, 10)

In [ ]:
#Create new variable holding final data. 
dfFixed = dfClean
dfFixed['date'] = pd.to_datetime(dfFixed['date']) #using pandas .to_datetime() to convert entire column into datetime object
dfFixed.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1188 entries, 0 to 1299
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   trip_id          1188 non-null   int64         
 1   date             1188 non-null   datetime64[ns]
 2   route            1188 non-null   object        
 3   vehicle_type     1188 non-null   object        
 4   time_of_day      1188 non-null   object        
 5   passenger_count  1188 non-null   float64       
 6   delay_minutes    1188 non-null   float64       
 7   fare_type        1188 non-null   object        
 8   payment_method   1188 non-null   object        
 9   day_type         1188 non-null   object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(6)
memory usage: 102.1+ KB


C:\Users\norma\AppData\Local\Temp\ipykernel_17988\2334423170.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfFixed['date'] = pd.to_datetime(dfFixed['date'])


In [74]:
dfFixed.head()
dfFixed.shape

(1188, 10)